### PR TITLE
feat(eslint-plugin): [no-floating-promise] add option to ignore IIFEs

### LINKED ISSUE #647: [no-floating-promises] Should not report on async IIFEs
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/no-floating-promises": "error"
  }
}
```

```TS
(async () => {
	await something();
})();
```

**Expected Result**

I expected it not to report as it's an intentional action. It doesn't catch a mistake.

If you're not willing to add this exception by default, at least add an option to ignore async IIFEs.

**Actual Result**

It reports as a floating promise, which is technically correct, but not wanted behavior, at least from my perspective.

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `1.11.0` |
| `@typescript-eslint/parser`        | `1.11.0` |
| `TypeScript`                       | `3.5.1` |
| `ESLint`                           | `5.16.0` |
| `node`                             | `8.16.0` |
| `npm`                              | `6.9.0` |

### PR DESCRIPTION
**Pull request by anikethsaha**
fixes #647 
added a no check for iife 
- created a new function to detect iife `isIife` (please do suggest if the logic there doesn't cover all cases)
- added tests

### CODE REVIEW

**Comment by bradzacher** [comment]
could you please gate this functionality behind an option, as per my comment: https://github.com/typescript-eslint/typescript-eslint/issues/647#issuecomment-505929666

I want end-users to opt-in to this functionality so that they're consciously making the decision for a little unsafety from the rule.

**Comment by anikethsaha** [comment]
> could you please gate this functionality behind an option, as per my comment: #647 (comment)
> 
> I want end-users to opt-in to this functionality so that they're consciously making the decision for a little unsafety from the rule.

ok cool, also it should only check for async iife right ? not just iife ?

**Comment by bradzacher** [comment]
it should check both - it should just be looking for an IIFE that returns a promise.

**Comment by anikethsaha** [comment]
> it should check both - it should just be looking for an IIFE that returns a promise.

make sense , ok creating an options name `ignoreIife`

**Comment by anikethsaha** [comment]
@bradzacher I think we need to change the selector for AST as well, cause

currently in `ExpressionStatement`

I have this check

```js
if (options.ignoreIife && isAsyncIife(node)) {
  return;
}
```

And, here is the definition of `isAsyncIife`

```typescript
    function isAsyncIife(node: TSESTree.ExpressionStatement): Boolean {
      if(node?.expression.type === AST_NODE_TYPES.AwaitExpression){
        return true
      }
      if (node?.expression.type === AST_NODE_TYPES.CallExpression) {
        if (possibleIifeCalleType.has(node?.expression?.callee.type)) {
          if (node?.expression?.callee?.hasAsync) {
            return true ;
          }
          return node?.expression?.callee?.body?.body.some(hasPromise);
        }
      }

      return false;
    }
```

And here is `hasPromise`

```typescript
    function hasPromise( node: TSESTree.ExpressionStatement | TSESTree.ReturnStatement): Boolean {
      // has promise statement
      if (node?.expression.type === "NewExpression" && node?.expression?.callee.name === "Promise") {
        return true;
      }
      if (
        // returning promise
        node.type === AST_NODE_TYPES.ReturnStatement &&
        node?.argument.type === AST_NODE_TYPES.NewExpression &&
        node?.argument?.callee.name === "Promise"
      ) {
        return true;
      }
      return false
    }
```

It looks like when I give input like this

```js
//  options : [{ ignoreIife : true }]
(async function() {
  // error cause of this
  await res(1);
})();
```

still the there is a floating error.

I did try changing the selectors

```typescript
{
["ExpressionStatement > :not(ExpressionStatement )"](node){
  check(node)
},
"ExpressionStatement > ExpressionStatement "(node){
if(node?.expression.type === AST_NODE_TYPES.AwaitExpression){
        return true
      }
check(node)
}

}

```

Still no luck

**Comment by bradzacher** [comment]
the code you've got currently in the PR doesn't look too far off.

I'd expect something along the lines of:
```ts
ExpressionStatement(node): void {
  if (options.ignoreIIFE && isIIFE(node)) {
    return;
  }

  // old code
}

function isIIFE(node: TSESTree.ExpressionStatement): boolean {
  const expr = node.expression;
  if (expr.type !== AST_NODE_TYPES.CallExpression) {
    return false;
  }

  return (
    expr.callee.type === AST_NODE_TYPES.ArrowFunctionExpression ||
    expr.callee.type === AST_NODE_TYPES.FunctionExpression
  );
}
```

To clarify what this option should do:

These cases should error with `ignoreIIFE: false` (the default)
They should **not** error with `ignoreIIFE: true`
```ts
(function foo() { return Promise.resolve() })();
(async function foo() { return Promise.resolve() })();
(async () => { return Promise.resolve() })();
```

However, setting the option to true should not ignore the contents of an IIFE. I.e. this should still error with `ignoreIIFE: true`
```ts
(async () => {
  Promise.resolve(); // error - floating promise
})();
// no other lint error
```

**Comment by anikethsaha** [comment]
@bradzacher should I add custom type for `isAsyncIife` method as I added `TSESTree.ExpressionStatement | TSESTree.ArrowFunctionExpression | TSESTree.TSTypeAnnotation` type to `node ` but still typecheck is failing. 
also, i tried `node?.expression.typeAnnotation.type`and still it failed

File: packages/eslint-plugin/docs/rules/no-floating-promises.md [hunk 1]
```diff
@@ -73,6 +75,33 @@ void returnsPromise();
 void Promise.reject('value');
 ```
 
+### `ignoreIIFE`
+
+This allows to skip checking of async iife
+
+Examples of **correct** code for this rule with `{ ignoreIIFE: true }`:
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-floating-promises.md]:
```suggestion
Examples of **correct** code for this rule with `{ ignoreIIFE: false }`:
```

File: packages/eslint-plugin/docs/rules/no-floating-promises.md [hunk 2]
```diff
@@ -73,6 +75,33 @@ void returnsPromise();
...
+
+```ts
+(async function() {
+  await res(1);
+})();
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-floating-promises.md]:
```suggestion
await (async function() {
  await res(1);
})();
```

File: packages/eslint-plugin/docs/rules/no-floating-promises.md [hunk 3]
```diff
@@ -73,6 +75,33 @@ void returnsPromise();
...
+      await res(1);
+    })();
+  });
+```
+
+Examples of **incorrect** code for this rule with `{ ignoreIIFE: true }`:
+
+```ts
+(async function() {
+  Promise.resolve();
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-floating-promises.md]:
```suggestion
  await Promise.resolve();
```

File: packages/eslint-plugin/src/rules/no-floating-promises.ts [hunk 4]
```diff
@@ -80,6 +91,20 @@ export default util.createRule<Options, MessageId>({
       },
     };
 
+    function isAsyncIife(
+      node: TSESTree.ExpressionStatement | TSESTree.ArrowFunctionExpression,
```
Comment (bradzacher) [inline packages/eslint-plugin/src/rules/no-floating-promises.ts]:
```suggestion
      node: TSESTree.ExpressionStatement,
```

File: packages/eslint-plugin/docs/rules/no-floating-promises.md (hunk 1 above)
Comment (anikethsaha) [reply packages/eslint-plugin/docs/rules/no-floating-promises.md]: I have kept `ignoreIIFE` false by default. 
so we need to add docs example when it is `true`, right !

File: packages/eslint-plugin/src/rules/no-floating-promises.ts (hunk 4 above)
Comment (anikethsaha) [reply packages/eslint-plugin/src/rules/no-floating-promises.ts]: `ExpressionStatement` doesnt have `ArrowFunctionExpression` expression type so its causing no overlap issue

image

Comment (bradzacher) [reply packages/eslint-plugin/src/rules/no-floating-promises.ts]: That's just an incorrect type definition.
You can add it to the type definition here
https://github.com/typescript-eslint/typescript-eslint/blob/133f622f38a286eac45288a9789d2ee529d5e582/packages/typescript-estree/src/ts-estree/ts-estree.ts#L384-L396

File: packages/eslint-plugin/docs/rules/no-floating-promises.md (hunk 1 above)
Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/no-floating-promises.md]: yeah sure. Don't need the false example.
Though I would avoid splitting the code block in two

eg do

---

Examples of **correct** code for this rule with `{ ignoreIIFE: true }`:

```ts

<examples>

```

<<end section>>

Comment (anikethsaha) [reply packages/eslint-plugin/docs/rules/no-floating-promises.md]: didnt get this .

You meant like this ?

**Before**
image


**After**
image

File: packages/eslint-plugin/docs/rules/no-floating-promises.md [hunk 5]
```diff
@@ -73,6 +75,33 @@ void returnsPromise();
...
 ```
 
+### `ignoreIIFE`
+
+This allows to skip checking of async iife
+
+Examples of **correct** code for this rule with `{ ignoreIIFE: true }`:
+
+```ts
+await(async function() {
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-floating-promises.md]: nit

```suggestion
await (async function() {
```

File: packages/eslint-plugin/tests/rules/no-floating-promises.test.ts [hunk 6]
```diff
@@ -250,7 +250,53 @@ async function test() {
...
+        })
+        `,
+      options: [{ ignoreIIFE: true }],
+    },
+    {
+      code: `(async function(){
+              await res(1)
+            })()`,
+      options: [{ ignoreIIFE: true }],
+    },
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/no-floating-promises.test.ts]: nit - formatting
```suggestion
      code: `
        (async () => {
          await something();
        })();
      `,
      options: [{ ignoreIIFE: true }],
    },
    {
      code: `
        (async () => {
          something();
        })();
      `,
      options: [{ ignoreIIFE: true }],
    },
    {
      code: '(async function foo(){})();',
      options: [{ ignoreIIFE: true }],
    },
    {
      code: `
        function foo(){
          (async function bar() {})()
        }
      `,
      options: [{ ignoreIIFE: true }],
    },
    {
      code: `
        const foo = () => new Promise((res) => {
          (async function(){
              await res(1)
          })()
        })
      `,
      options: [{ ignoreIIFE: true }],
    },
    {
      code: `
        (async function(){
              await res(1)
        })()
      `,
      options: [{ ignoreIIFE: true }],
    },
```

File: packages/eslint-plugin/docs/rules/no-floating-promises.md [hunk 7]
```diff
@@ -73,6 +75,33 @@ void returnsPromise();
...
+    (async function() {
+      await res(1);
+    })();
+  });
+```
+
+Examples of **incorrect** code for this rule with `{ ignoreIIFE: true }`:
+
+```ts
+await(async function() {
```
Comment (bradzacher) [inline packages/eslint-plugin/docs/rules/no-floating-promises.md]: This example is supposed to be a floating IIFE, so no await

```suggestion
(async function() {
```

File: packages/eslint-plugin/tests/rules/no-floating-promises.test.ts [hunk 8]
```diff
@@ -642,5 +688,119 @@ async function test() {
...
+        {
+          line: 5,
+          messageId: 'floating',
+        },
+        {
+          line: 6,
+          messageId: 'floating',
+        },
+      ],
+    },
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/no-floating-promises.test.ts]: nit - formatting
```suggestion
    {
      code: `
        (async () => {
          await something();
        })();
      `,
      errors: [
        {
          line: 2,
          messageId: 'floating',
        },
      ],
    },
    {
      code: `
        (async () => {
          something();
        })();
      `,
      errors: [
        {
          line: 2,
          messageId: 'floating',
        },
      ],
    },
    {
      code: '(async function foo(){})();',
      errors: [
        {
          line: 1,
          messageId: 'floating',
        },
      ],
    },
    {
      code: `
        function foo(){
          (async function bar() {})()
        }
      `,
      errors: [
        {
          line: 3,
          messageId: 'floating',
        },
      ],
    },
    {
      code: `
        const foo = () => new Promise((res) => {
          (async function(){
            await res(1)
          })()
        })
      `,
      errors: [
        {
          line: 2,
          messageId: 'floating',
        },
      ],
    },
    {
      code: `
        (async function(){
          await res(1)
        })()
      `,
      errors: [
        {
          line: 1,
          messageId: 'floating',
        },
      ],
    },
    {
      code: `
        (async function(){
          Promise.resolve();
        })()
      `,
      options: [{ ignoreIIFE: true }],
      errors: [
        {
          line: 2,
          messageId: 'floating',
        },
      ],
    },
    {
      code: `
        (async function(){
          const promiseIntersection: Promise<number> & number;
          promiseIntersection;
          promiseIntersection.then(() => {})
          promiseIntersection.catch();
          promiseIntersection.finally();
        })()
      `,
      options: [{ ignoreIIFE: true }],
      errors: [
        {
          line: 3,
          messageId: 'floating',
        },
        {
          line: 4,
          messageId: 'floating',
        },
        {
          line: 5,
          messageId: 'floating',
        },
        {
          line: 6,
          messageId: 'floating',
        },
      ],
    },
```

**Comment by bradzacher** [comment]
make sure you run prettier over your code

https://github.com/typescript-eslint/typescript-eslint/blob/master/CONTRIBUTING.md#validating-your-changes

**Comment by anikethsaha** [comment]
same issue here as well 

image

File: packages/eslint-plugin/docs/rules/no-floating-promises.md (hunk 5 above)
Comment (JoshuaKGoldberg) [reply packages/eslint-plugin/docs/rules/no-floating-promises.md:86]: #8228

**Pull request merged**